File: rust/src/ixgbe/regs.rs
#![allow(non_snake_case)]
#![allow(clippy::eq_op)]
#![allow(clippy::erasing_op)]

use crate::lifed::LifedSlice;

// Regs are usize so they can index a slice without ceremony

pub const CTRL: usize = 0x00000 / 4;
pub mod CTRL_ {
    pub const MASTER_DISABLE: u32 = 1 << 2;
    pub const RST: u32 = 1 << 26;
}

pub const CTRLEXT: usize = 0x00018 / 4;
pub mod CTRLEXT_ {
    pub const NSDIS: u32 = 1 << 16;
}

pub const fn DCARXCTRL(n: usize) -> usize {
    if n <= 63 {
        0x0100C / 4 + 0x10 * n
    } else {
        0x0D00C + 0x40 * (n - 64)
    }
}
pub mod DCARXCTRL_ {
    pub const UNKNOWN: u32 = 1 << 12;
}

pub const fn DCATXCTRL(n: usize) -> usize {
    0x0600C / 4 + 0x10 * n
}
pub mod DCATXCTRL_ {
    pub const TX_DESC_WB_RO_EN: u32 = 1 << 11;
}

pub const DMATXCTL: usize = 0x04A80 / 4;
pub mod DMATXCTL_ {
    pub const TE: u32 = 1 << 0;
}

pub const DTXMXSZRQ: usize = 0x08100 / 4;
pub mod DTXMXSZRQ_ {
    pub const MAX_BYTES_NUM_REQ: u32 = 0b0111_1111_1111;
}

pub const EEC: usize = 0x10010 / 4;
pub mod EEC_ {
    pub const EE_PRES: u32 = 1 << 8;
    pub const AUTO_RD: u32 = 1 << 9;
}

pub const fn EIMC(n: usize) -> usize {
    if n == 0 {
        0x00888 / 4
    } else {
        0x00AB0 / 4 + (n - 1)
    }
}

pub const FCCFG: usize = 0x03D00 / 4;
pub mod FCCFG_ {
    pub const TFCE: u32 = 0b0001_1000;
}

pub const fn FCRTH(n: usize) -> usize {
    (0x03260 + 4 * n) / 4
}
pub mod FCRTH_ {
    pub const RTH: u32 = 0b0111_1111_1111_1110_0000;
}

pub const FCTRL: usize = 0x05080 / 4;
pub mod FCTRL_ {
    pub const MPE: u32 = 1 << 8;
    pub const UPE: u32 = 1 << 9;
}

pub const fn FTQF(n: usize) -> usize {
    0x0E600 / 4 + n
}
pub mod FTQF_ {
    pub const QUEUE_ENABLE: u32 = 1 << 31;
}

pub const FWSM: usize = 0x10148 / 4;
pub mod FWSM_ {
    pub const EXT_ERR_IND: u32 = 0b0001_1111_1000_0000_0000_0000_0000;
}

pub const GCREXT: usize = 0x11050 / 4;
pub mod GCREXT_ {
    pub const BUFFERS_CLEAR_FUNC: u32 = 1 << 30;
}

pub const HLREG0: usize = 0x04240 / 4;
pub mod HLREG0_ {
    pub const LPBK: u32 = 1 << 15;
}

pub const MFLCN: usize = 0x04294 / 4;
pub mod MFLCN_ {
    pub const RFCE: u32 = 1 << 3;
}

pub const fn MPSAR(n: usize) -> usize {
    0x0A600 / 4 + n
}

pub const fn MTA(n: usize) -> usize {
    0x05200 / 4 + n
}

pub const fn PFUTA(n: usize) -> usize {
    0x0F400 / 4 + n
}

pub const fn PFVLVF(n: usize) -> usize {
    0x0F100 / 4 + n
}

pub const fn PFVLVFB(n: usize) -> usize {
    0x0F200 / 4 + n
}

pub const RDRXCTL: usize = 0x02F00 / 4;
pub mod RDRXCTL_ {
    pub const CRC_STRIP: u32 = 1 << 1;
    pub const DMAIDONE: u32 = 1 << 3;
    pub const RSCFRSTSIZE: u32 = 0b0011_1110_0000_0000_0000_0000;
    pub const RSCACKC: u32 = 1 << 25;
    pub const FCOE_WRFIX: u32 = 1 << 26;
}

pub const RTTDCS: usize = 0x04900 / 4;
pub mod RTTDCS_ {
    pub const ARBDIS: u32 = 1 << 6;
}

pub const RTTDQSEL: usize = 0x04904 / 4;

pub const RTTDT1C: usize = 0x04908 / 4;

pub const RXCTRL: usize = 0x03000 / 4;
pub mod RXCTRL_ {
    pub const RXEN: u32 = 1 << 0;
}

pub const fn RXPBSIZE(n: usize) -> usize {
    0x03C00 / 4 + n
}

pub const SECRXCTRL: usize = 0x08D00 / 4;
pub mod SECRXCTRL_ {
    pub const RX_DIS: u32 = 1 << 1;
}

pub const SECRXSTAT: usize = 0x08D04 / 4;
pub mod SECRXSTAT_ {
    pub const SECRX_RDY: u32 = 1 << 0;
}

pub const fn SRRCTL(n: usize) -> usize {
    if n <= 63 {
        0x01014 / 4 + 0x10 * n
    } else {
        0x0D014 / 4 + 0x10 * (n - 64)
    }
}
pub mod SRRCTL_ {
    pub const BSIZEPACKET: u32 = 0b0001_1111;
    pub const DROP_EN: u32 = 1 << 28;
}

pub const STATUS: usize = 0x00008 / 4;
pub mod STATUS_ {
    pub const PCIE_MASTER_ENABLE_STATUS: u32 = 1 << 19;
}

pub const fn RDBAL(n: usize) -> usize {
    if n <= 63 {
        0x01000 / 4 + 0x10 * n
    } else {
        0x0D000 / 4 + 0x10 * (n - 64)
    }
}

pub const fn RDBAH(n: usize) -> usize {
    if n <= 63 {
        0x01004 / 4 + 0x10 * n
    } else {
        0x0D004 / 4 + 0x10 * (n - 64)
    }
}

pub const fn RDLEN(n: usize) -> usize {
    if n <= 63 {
        0x01008 / 4 + 0x10 * n
    } else {
        0x0D008 / 4 + 0x10 * (n - 64)
    }
}

pub const fn RDT(n: usize) -> usize {
    if n <= 63 {
        0x01018 / 4 + 0x10 * n
    } else {
        0x0D018 / 4 + 0x10 * (n - 64)
    }
}

pub const fn RXDCTL(n: usize) -> usize {
    if n <= 63 {
        0x01028 / 4 + 0x10 * n
    } else {
        0x0D028 / 4 + 0x10 * (n - 64)
    }
}
pub mod RXDCTL_ {
    pub const ENABLE: u32 = 1 << 25;
}

pub const fn TDBAH(n: usize) -> usize {
    0x06004 / 4 + 0x10 * n
}

pub const fn TDBAL(n: usize) -> usize {
    0x06000 / 4 + 0x10 * n
}

pub const fn TDLEN(n: usize) -> usize {
    0x06008 / 4 + 0x10 * n
}

pub const fn TDT(n: usize) -> usize {
    0x06018 / 4 + 0x10 * n
}

pub const fn TDWBAH(n: usize) -> usize {
    0x0603C / 4 + 0x10 * n
}

pub const fn TDWBAL(n: usize) -> usize {
    0x06038 / 4 + 0x10 * n
}

pub const fn TXDCTL(n: usize) -> usize {
    0x06028 / 4 + 0x10 * n
}
pub mod TXDCTL_ {
    pub const PTHRESH: u32 = 0b0111_1111;
    pub const HTHRESH: u32 = 0b0111_1111_0000_0000;
    pub const ENABLE: u32 = 1 << 25;
}

pub const fn TXPBSIZE(n: usize) -> usize {
    0x0CC00 / 4 + n
}

pub const fn TXPBTHRESH(n: usize) -> usize {
    0x04950 / 4 + n
}
pub mod TXPBTHRESH_ {
    pub const THRESH: u32 = 0b11_1111_1111;
}

pub fn read(buffer: LifedSlice<'_, u32>, reg: usize) -> u32 {
    u32::from_le(buffer.index(reg).read_volatile())
}

pub fn read_field(buffer: LifedSlice<'_, u32>, reg: usize, field: u32) -> u32 {
    let value = read(buffer, reg);
    let shift = field.trailing_zeros();
    (value & field) >> shift
}

pub fn write(buffer: LifedSlice<'_, u32>, reg: usize, value: u32) {
    buffer.index(reg).write_volatile(u32::to_le(value));
}

pub fn write_field(buffer: LifedSlice<'_, u32>, reg: usize, field: u32, field_value: u32) {
    let old_value = read(buffer, reg);
    let shift = field.trailing_zeros();
    let new_value = (old_value & !field) | (field_value << shift);
    write(buffer, reg, new_value);
}

pub fn clear(buffer: LifedSlice<'_, u32>, reg: usize) {
    write(buffer, reg, 0);
}

pub fn clear_field(buffer: LifedSlice<'_, u32>, reg: usize, field: u32) {
    write_field(buffer, reg, field, 0);
}

pub fn set_field(buffer: LifedSlice<'_, u32>, reg: usize, field: u32) {
    let old_value = read(buffer, reg);
    let new_value = old_value | field;
    write(buffer, reg, new_value);
}

pub fn is_field_cleared(buffer: LifedSlice<'_, u32>, reg: usize, field: u32) -> bool {
    read_field(buffer, reg, field) == 0
}

File: csharp/TinyNF/Ixgbe/Regs.cs
using System;
using System.Numerics;
using System.Threading;
using TinyNF.Environment;

namespace TinyNF.Ixgbe
{
    internal static class Regs
    {
        public const uint CTRL = 0x00000u / 4u;
        public static class CTRL_
        {
            public const uint MASTER_DISABLE = 1 << 2;
            public const uint RST = 1 << 26;
        }

        public const uint CTRLEXT = 0x00018u / 4u;
        public static class CTRLEXT_
        {
            public const uint NSDIS = 1 << 16;
        }

        public static uint DCARXCTRL(uint n) => n <= 63u ? (0x0100Cu / 4u + 0x10u * n) : (0x0D00Cu / 4u + 0x10u * (n - 64u));
        public static class DCARXCTRL_
        {
            public const uint UNKNOWN = 1 << 12;
        }

        public static uint DCATXCTRL(uint n) => 0x0600Cu / 4u + 0x10u * n;
        public static class DCATXCTRL_
        {
            public const uint TX_DESC_WB_RO_EN = 1 << 11;
        }

        public const uint DMATXCTL = 0x04A80u / 4u;
        public static class DMATXCTL_
        {
            public const uint TE = 1 << 0;
        }

        public const uint DTXMXSZRQ = 0x08100u / 4u;
        public static class DTXMXSZRQ_
        {
            public const uint MAX_BYTES_NUM_REQ = 0b0111_1111_1111;
        }

        public const uint EEC = 0x10010u / 4u;
        public static class EEC_
        {
            public const uint EE_PRES = 1 << 8;
            public const uint AUTO_RD = 1 << 9;
        }

        public static uint EIMC(uint n) => n == 0 ? (0x00888u / 4u) : (0x00AB0u / 4u + (n - 1u));

        public const uint FCCFG = 0x03D00u / 4u;
        public static class FCCFG_
        {
            public const uint TFCE = 0b0001_1000;
        }

        public static uint FCRTH(uint n) => 0x03260u / 4u + n;
        public static class FCRTH_
        {
            public const uint RTH = 0b0111_1111_1111_1110_0000;
        }

        public const uint FCTRL = 0x05080u / 4u;
        public static class FCTRL_
        {
            public const uint MPE = 1 << 8;
            public const uint UPE = 1 << 9;
        }

        public static uint FTQF(uint n) => 0x0E600u / 4u + n;
        public static class FTQF_
        {
            public const uint QUEUE_ENABLE = 1u << 31;
        }

        public const uint FWSM = 0x10148u / 4u;
        public static class FWSM_
        {
            public const uint EXT_ERR_IND = 0b0001_1111_1000_0000_0000_0000_0000;
        }

        public const uint GCREXT = 0x11050u / 4u;
        public static class GCREXT_
        {
            public const uint BUFFERS_CLEAR_FUNC = 1 << 30;
        }

        public const uint HLREG0 = 0x04240u / 4u;
        public static class HLREG0_
        {
            public const uint LPBK = 1 << 15;
        }

        public const uint MFLCN = 0x04294u / 4u;
        public static class MFLCN_
        {
            public const uint RFCE = 1 << 3;
        }

        public static uint MPSAR(uint n) => 0x0A600u / 4u + n;

        public static uint MTA(uint n) => 0x05200u / 4u + n;

        public static uint PFUTA(uint n) => 0x0F400u / 4u + n;

        public static uint PFVLVF(uint n) => 0x0F100u / 4u + n;

        public static uint PFVLVFB(uint n) => 0x0F200u / 4u + n;

        public static uint RDBAH(uint n) => n <= 63u ? (0x01004u / 4u + 0x10u * n) : (0x0D004u / 4u + 0x10u * (n - 64u));

        public static uint RDBAL(uint n) => n <= 63u ? (0x01000u / 4u + 0x10u * n) : (0x0D000u / 4u + 0x10u * (n - 64u));

        public static uint RDLEN(uint n) => n <= 63u ? (0x01008u / 4u + 0x10u * n) : (0x0D008u / 4u + 0x10u * (n - 64u));

        public const uint RDRXCTL = 0x02F00u / 4u;
        public static class RDRXCTL_
        {
            public const uint CRC_STRIP = 1 << 1;
            public const uint DMAIDONE = 1 << 3;
            public const uint RSCFRSTSIZE = 0b0011_1110_0000_0000_0000_0000;
            public const uint RSCACKC = 1 << 25;
            public const uint FCOE_WRFIX = 1 << 26;
        }

        public static uint RDT(uint n) => n <= 63u ? (0x01018u / 4u + 0x10u * n) : (0x0D018u / 4u + 0x10u * (n - 64u));

        public const uint RTTDCS = 0x04900u / 4u;
        public static class RTTDCS_
        {
            public const uint ARBDIS = 1 << 6;
        }

        public const uint RTTDQSEL = 0x04904u / 4u;

        public const uint RTTDT1C = 0x04908u / 4u;

        public const uint RXCTRL = 0x03000u / 4u;
        public static class RXCTRL_
        {
            public const uint RXEN = 1 << 0;
        }

        public static uint RXDCTL(uint n) => n <= 63u ? (0x01028u / 4u + 0x10u * n) : (0x0D028u / 4u + 0x10u * (n - 64u));
        public static class RXDCTL_
        {
            public const uint ENABLE = 1 << 25;
        }

        public static uint RXPBSIZE(uint n) => 0x03C00u / 4u + n;

        public const uint SECRXCTRL = 0x08D00u / 4u;
        public static class SECRXCTRL_
        {
            public const uint RX_DIS = 1 << 1;
        }

        public const uint SECRXSTAT = 0x08D04u / 4u;
        public static class SECRXSTAT_
        {
            public const uint SECRX_RDY = 1 << 0;
        }

        public static uint SRRCTL(uint n) => n <= 63u ? (0x01014u / 4u + 0x10u * n) : (0x0D014u / 4u + 0x10u * (n - 64u));
        public static class SRRCTL_
        {
            public const uint BSIZEPACKET = 0b0001_1111;
            public const uint DROP_EN = 1 << 28;
        }

        public const uint STATUS = 0x00008u / 4u;
        public static class STATUS_
        {
            public const uint PCIE_MASTER_ENABLE_STATUS = 1 << 19;
        }

        public static uint TDBAH(uint n) => 0x06004u / 4u + 0x10u * n;

        public static uint TDBAL(uint n) => 0x06000u / 4u + 0x10u * n;

        public static uint TDLEN(uint n) => 0x06008u / 4u + 0x10u * n;

        public static uint TDT(uint n) => 0x06018u / 4u + 0x10u * n;

        public static uint TDWBAH(uint n) => 0x0603Cu / 4u + 0x10u * n;

        public static uint TDWBAL(uint n) => 0x06038u / 4u + 0x10u * n;

        public static uint TXDCTL(uint n) => 0x06028u / 4u + 0x10u * n;
        public static class TXDCTL_
        {
            public const uint PTHRESH = 0b0111_1111;
            public const uint HTHRESH = 0b0111_1111_0000_0000;
            public const uint ENABLE = 1 << 25;
        }

        public static uint TXPBSIZE(uint n) => 0x0CC00u / 4u + n;

        public static uint TXPBTHRESH(uint n) => 0x04950u / 4u + n;
        public static class TXPBTHRESH_
        {
            public const uint THRESH = 0b11_1111_1111;
        }


        public static uint Read(Memory<uint> buffer, uint reg)
        {
            return Endianness.FromLittle(Volatile.Read(ref buffer.Span[(int)reg]));
        }

        public static uint ReadField(Memory<uint> buffer, uint reg, uint field)
        {
            uint value = Read(buffer, reg);
            int shift = BitOperations.TrailingZeroCount(field);
            return (value & field) >> shift;
        }

        public static void Write(Memory<uint> buffer, uint reg, uint value)
        {
            Volatile.Write(ref buffer.Span[(int)reg], Endianness.ToLittle(value));
        }

        public static void WriteField(Memory<uint> buffer, uint reg, uint field, uint fieldValue)
        {
            uint oldValue = Read(buffer, reg);
            int shift = BitOperations.TrailingZeroCount(field);
            uint newValue = (oldValue & ~field) | (fieldValue << shift);
            Write(buffer, reg, newValue);
        }

        public static void Clear(Memory<uint> buffer, uint reg)
        {
            Write(buffer, reg, 0);
        }

        public static void ClearField(Memory<uint> buffer, uint reg, uint field)
        {
            WriteField(buffer, reg, field, 0);
        }

        public static void SetField(Memory<uint> buffer, uint reg, uint field)
        {
            uint oldValue = Read(buffer, reg);
            uint newValue = oldValue | field;
            Write(buffer, reg, newValue);
        }

        public static bool IsFieldCleared(Memory<uint> buffer, uint reg, uint field)
        {
            return ReadField(buffer, reg, field) == 0;
        }
    }
}
